### PR TITLE
tommy/works_backend_pagination

### PR DESCRIPTION
**Pull request by tommypoa**
## Feature Name: Backend Pagination for All Works
Backend pagination implemented for all works page.

### Related PRs
None.

### Migrations
None.

### Tests Performed, Edge Cases
Tested on admin, buyer and artist logged in status, as well as ensuring it works with filtering.

CC: @by-co

### CODE REVIEW

File: app/controllers/api/works_controller.rb [hunk 1]
```diff
@@ -97,22 +97,31 @@ def destroy
...
+      works: ActiveModel::Serializer::CollectionSerializer.new(filtered_works, each_serializer: WorkSerializer),
+      work_count: work_count
+    }
   end
 
   def filtered_artist_hidden
-    filtered_works = Work.joins(:artist).where("artists.hidden=false").select{|work| work.hidden==false}
-    render json: filtered_works,
-      each_serializer: WorkSerializer
+    filtered_works = Work.joins(:artist).where("artists.hidden" => false, "works.hidden" => false).page(params[:page])
```
Comment (by-co) [inline app/controllers/api/works_controller.rb]: Might be a little cleaner to save `Work.joins(:artist).where("artists.hidden" => false, "works.hidden" => false)` in a local variable?

File: app/controllers/api/works_controller.rb [hunk 2]
```diff
@@ -97,22 +97,31 @@ def destroy
...
 
   def index
-    works = Work.all
-    render json: works,
-      each_serializer: WorkSerializer
+    work_count = Work.count
+    works = Work.page(params[:page])
+    render json: {
+      works: ActiveModel::Serializer::CollectionSerializer.new(works, each_serializer: WorkSerializer),
+      work_count: work_count
```
Comment (aivantg) [inline app/controllers/api/works_controller.rb]: why not just replace work_count with Work.count? just to be clean

File: app/controllers/api/works_controller.rb [hunk 3]
```diff
@@ -97,22 +97,31 @@ def destroy
...
+    }
   end
 
   def filtered_works
     parsed_query = CGI.parse(params[:search_params])
-    filtered_works = params[:search_params] == "" ?  Work.all : Work.where(parsed_query)
-    render json: filtered_works,
-      each_serializer: WorkSerializer
+    filtered_works = params[:search_params] == "" ?  Work.all.page(params[:page]) : Work.where(parsed_query).page(params[:page])
```
Comment (aivantg) [inline app/controllers/api/works_controller.rb]: don't repeat these sql queries! call `Work.where` just once and use the variable twice :)

File: app/controllers/api/works_controller.rb [hunk 4]
```diff
@@ -97,22 +97,31 @@ def destroy
...
+      work_count: work_count
+    }
   end
 
   def filtered_artist_hidden
-    filtered_works = Work.joins(:artist).where("artists.hidden=false").select{|work| work.hidden==false}
-    render json: filtered_works,
-      each_serializer: WorkSerializer
+    filtered_works = Work.joins(:artist).where("artists.hidden" => false, "works.hidden" => false).page(params[:page])
+    work_count =  Work.joins(:artist).where("artists.hidden" => false, "works.hidden" => false).count
```
Comment (aivantg) [inline app/controllers/api/works_controller.rb]: Don't make this query twice!

File: app/javascript/components/works/Works.jsx [hunk 5]
```diff
@@ -21,54 +21,57 @@ class Works extends React.Component {
...
       const [works_response, filters_response] = response;
       this.setState({
-        works: works_response,
+        works: works_response.works,
         filters: filters_response,
         isLoading: false,
-        pageCount: Math.ceil(works_response.length / perPage),
-        workStartIndex: 0,
-        workEndIndex: perPage
+        pageCount: Math.ceil(works_response.work_count / perPage)
```
Comment (aivantg) [inline app/javascript/components/works/Works.jsx]: perPage should not be hardcoded into the front end! Send it back as part of the pagination response as well so there's only 1 source of truth.

File: app/controllers/api/works_controller.rb [hunk 6]
```diff
@@ -97,22 +97,31 @@ def destroy
...
 
   def filtered_works
     parsed_query = CGI.parse(params[:search_params])
-    filtered_works = params[:search_params] == "" ?  Work.all : Work.where(parsed_query)
-    render json: filtered_works,
-      each_serializer: WorkSerializer
+    filtered_works = params[:search_params] == "" ?  Work.all.page(params[:page]) : Work.where(parsed_query).page(params[:page])
+    work_count = params[:search_params] == "" ?  Work.count : Work.where(parsed_query).count
+    render json: {
+      works: ActiveModel::Serializer::CollectionSerializer.new(filtered_works, each_serializer: WorkSerializer),
```
[the remark this answers is not in the document]
Comment (tommypoa) [reply app/controllers/api/works_controller.rb]: Tested and it seems necessary!

**Comment by by-co** [comment]
Ran a test on Network to check if this is making a difference - notice "Disable cache" is checked

Old method fetches 18KB (really like nothing, but our seeded DB only has like 13 works)
frontend pagination

Fetching 6 works at a time reduces that bundle!
tommy:backend_pagination

**Comment by by-co** [comment]
Looks good to me, @tommypoa please resolve merge conflicts and wait for Aivant to review 👍
